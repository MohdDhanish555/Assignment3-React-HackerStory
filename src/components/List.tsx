import {
  TableContainer,
  Paper,
  Table,
  TableHead,
  TableRow,
  TableCell,
  TableBody,
} from "@mui/material";
import { styled } from "@mui/material/styles";
import { tableCellClasses } from "@mui/material/TableCell";
import { useNavigate } from "react-router-dom";
import { useTheme } from "@mui/material";
import HighlightOffIcon from "@mui/icons-material/HighlightOff";

const StyledTableCell = styled(TableCell)(({ theme }) => ({
  [`&.${tableCellClasses.head}`]: {
    backgroundColor: theme.palette.secondary.main,
    color: theme.palette.common.white,
    padding: ".8rem",
  },
}));

const StyledTableRow = styled(TableRow)(({ theme }) => ({
  "&:nth-of-type(odd)": {
    backgroundColor: theme.palette.action.hover,
  },
  // hide last border
  "&:last-child td, &:last-child th": {
    border: 0,
  },
}));

const List = ({ stories, setStories, lastHitElement }: any) => {
  const theme = useTheme();

  function handleDeleteClick(itemID: number) {
    setStories(stories?.filter((story: any) => story.objectID !== itemID));
  }

  const navigate = useNavigate();

  function handleTitleClick(itemID: number) {
    stories?.map((story: any) => {
      if (story.objectID === itemID)
        return navigate("/rawContent", { state : { story } });
    });
  }

  return (
    <TableContainer
      component={Paper}
      sx={{ maxWidth: "1400px", mt: "1rem", mb: "5rem" }}
    >
      <Table size="small" aria-label="simple table">
        <TableHead>
          <TableRow>
            <StyledTableCell sx={{ width: "350px" }}>Title</StyledTableCell>
            <StyledTableCell sx={{ width: "350px" }}>URL</StyledTableCell>
            <StyledTableCell sx={{ width: "210px" }}>Author</StyledTableCell>
            <StyledTableCell sx={{ width: "150px" }}>Comments</StyledTableCell>
            <StyledTableCell sx={{ width: "100px" }}>Action</StyledTableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {stories?.map((item: any, index: number) => {
            if (stories.length === index + 1) {
              return (
                <StyledTableRow
                  key={item.objectID}
                  ref={lastHitElement}
                  sx={{
                    "&:last-child td, &:last-child th": {
                      border: 0,
                    },
                  }}
                >
                  <TableCell
                    data-testid="cell-title1"
                    sx={{
                      maxWidth: "350px",
                      whiteSpace: "nowrap",
                      overflow: "hidden",
                      textOverflow: "ellipsis",
                      cursor: "pointer",
                    }}
                    onClick={() => handleTitleClick(item.objectID)}
                  >
                    {item.title}
                  </TableCell>
                  <TableCell
                    sx={{
                      maxWidth: "350px",
                      whiteSpace: "nowrap",
                      overflow: "hidden",
                      textOverflow: "ellipsis",
                    }}
                  >
                    {item.url}
                  </TableCell>
                  <TableCell
                    sx={{
                      minWidth: "210px",
                      whiteSpace: "nowrap",
                      overflow: "hidden",
                      textOverflow: "ellipsis",
                    }}
                  >
                    {item.author}
                  </TableCell>
                  <TableCell
                    sx={{
                      minWidth: "150px",
                      whiteSpace: "nowrap",
                      overflow: "hidden",
                      textOverflow: "ellipsis",
                    }}
                  >
                    {item.num_comments}
                  </TableCell>
                  <TableCell
                    data-testid="cell-delete1"
                    sx={{
                      minWidth: "100px",
                      whiteSpace: "nowrap",
                      overflow: "hidden",
                      textOverflow: "ellipsis",
                      color: theme.palette.primary.main,
                      cursor: "pointer",
                      display: "flex",
                      gap: ".5rem",
                    }}
                    onClick={() => handleDeleteClick(item.objectID)}
                  >
                    Delete
                    <HighlightOffIcon fontSize="small" />
                  </TableCell>
                </StyledTableRow>
              );
            } else {
              return (
                <StyledTableRow
                  key={item.objectID}
                  sx={{ "&:last-child td, &:last-child th": { border: 0 } }}
                >
                  <TableCell
                    data-testid="cell-title2"
                    sx={{
                      maxWidth: "350px",
                      whiteSpace: "nowrap",
                      overflow: "hidden",
                      textOverflow: "ellipsis",
                      cursor: "pointer",
                    }}
                    onClick={() => handleTitleClick(item.objectID)}
                  >
                    {item.title}
                  </TableCell>
                  <TableCell
                    sx={{
                      maxWidth: "350px",
                      whiteSpace: "nowrap",
                      overflow: "hidden",
                      textOverflow: "ellipsis",
                    }}
                  >
                    {item.url}
                  </TableCell>
                  <TableCell
                    sx={{
                      minWidth: "210px",
                      whiteSpace: "nowrap",
                      overflow: "hidden",
                      textOverflow: "ellipsis",
                    }}
                  >
                    {item.author}
                  </TableCell>
                  <TableCell
                    sx={{
                      minWidth: "150px",
                      whiteSpace: "nowrap",
                      overflow: "hidden",
                      textOverflow: "ellipsis",
                    }}
                  >
                    {item.num_comments}
                  </TableCell>
                  <TableCell
                    data-testid="cell-delete2"
                    sx={{
                      minWidth: "100px",
                      whiteSpace: "nowrap",
                      overflow: "hidden",
                      textOverflow: "ellipsis",
                      color: theme.palette.primary.main,
                      cursor: "pointer",
                      display: "flex",
                      gap: ".5rem",
                    }}
                    onClick={() => handleDeleteClick(item.objectID)}
                  >
                    Delete
                    <HighlightOffIcon fontSize="small" />
                  </TableCell>
                </StyledTableRow>
              );
            }
          })}
        </TableBody>
      </Table>
    </TableContainer>
  );
};

export default List;
