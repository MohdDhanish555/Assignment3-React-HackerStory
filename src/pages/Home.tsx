import { Backdrop, Card, CardMedia, Button, CardContent } from "@mui/material";
import { useState } from "react";
import { useNavigate } from "react-router-dom";
import paginationGif from "../assets/images/pagination-3.gif";
import scrollGif from "../assets/images/mouse-scroll.gif";

const Home = () => {
  const [open, setOpen] = useState(true);
    
  // HANDLING BACKDROP
  function handleButton1() {
    setOpen(false);
    navigate("/VariantOne");
  };

  function handleButton2() {
    setOpen(false);
    navigate("/VariantTwo");
  };

  const navigate = useNavigate();
  return (
    <div>
      <Backdrop
        sx={{
          color: "#fff",
          zIndex: (theme) => theme.zIndex.drawer + 1,
          gap: "10rem",
        }}
        open={open}
      >
        <Card raised sx={{ maxWidth: 340, bgcolor: "rgba(0,0,0,0.2)" }}>
          <CardMedia
            component="img"
            height="194"
            image={paginationGif}
            alt="Variant 1"
          />
          <CardContent
            sx={{
              display: "flex",
              justifyContent: "center",
            }}
          >
            <Button
              variant="outlined"
              data-testid="btn-V1"
              onClick={handleButton1}
            >
              Variant 1
            </Button>
          </CardContent>
        </Card>
        <Card
          raised
          sx={{
            maxWidth: 340,
            bgcolor: "rgba(0,0,0,0.2)",
          }}
        >
          <CardMedia
            component="img"
            height="194"
            image={scrollGif}
            alt="Variant 2"
          />
          <CardContent sx={{ display: "flex", justifyContent: "center" }}>
            <Button
              data-testid="btn-V2"
              variant="outlined"
              onClick={handleButton2}
            >
              Variant 2
            </Button>
          </CardContent>
        </Card>
      </Backdrop>
    </div>
  );
};

export default Home;
