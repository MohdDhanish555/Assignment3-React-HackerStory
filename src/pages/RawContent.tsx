import { Box, Button} from "@mui/material";
import { useLocation, useNavigate } from "react-router-dom";

const RawContent = () => {
  const { state }: any = useLocation();
  const newStory = JSON.stringify(state?.story, null, 4);
    const navigate = useNavigate();
    
  return (
    <>
      <Box display="flex" justifyContent="space-between">
      <pre>{newStory}</pre>;
        <Button
          sx={{
            alignSelf: "flex-start",
            margin : "1rem"
        }}
          variant="outlined"
          data-testid="btn-Raw"
          onClick={() => navigate(-1)}
        >
          Back
        </Button>
      </Box>
    </>
  );
};

export default RawContent;
