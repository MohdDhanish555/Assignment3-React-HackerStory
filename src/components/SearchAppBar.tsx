import { styled, alpha } from "@mui/material/styles";
import {
  AppBar,
  Box,
  Toolbar,
  Typography,
  InputBase,
  LinearProgress,
} from "@mui/material";
import SearchIcon from "@mui/icons-material/Search";
import { useContext } from "react";
import { AppContext } from "../App";
import { useTheme } from "@mui/material";

const Search = styled("div")(({ theme }) => ({
  position: "relative",
  borderRadius: "20px",
  backgroundColor: alpha(theme.palette.common.white, 0.15),
  "&:hover": {
    backgroundColor: alpha(theme.palette.common.white, 0.25),
  },
  marginLeft: 0,
  marginRight: "5rem",
  width: "100%",
  [theme.breakpoints.up("sm")]: {
    marginLeft: theme.spacing(1),
    width: "auto",
  },
}));

const SearchIconWrapper = styled("div")(({ theme }) => ({
  padding: theme.spacing(0, 2),
  height: "100%",
  position: "absolute",
  pointerEvents: "none",
  display: "flex",
  alignItems: "center",
  justifyContent: "center",
}));

const StyledInputBase = styled(InputBase)(({ theme }) => ({
  color: "inherit",
  "& .MuiInputBase-input": {
    padding: theme.spacing(1, 1, 1, 0),
    // vertical padding + font size from searchIcon
    paddingLeft: `calc(1em + ${theme.spacing(4)})`,
    transition: theme.transitions.create("width"),
    width: "100%",
    [theme.breakpoints.up("sm")]: {
      width: "20ch",
      "&:focus": {
        width: "40ch",
      },
    },
  },
}));

const SearchAppBar = () => {
  const [queryText, setQueryText] = useContext(AppContext);
const theme = useTheme();
  function handleChange(event: React.ChangeEvent<HTMLInputElement>) {
    setQueryText(event.target.value);
  }

  return (
    <Box sx={{ flexGrow: 1 }}>
      <AppBar
        position="static"
        sx={{ p: ".5rem 2rem", bgcolor: theme.palette.secondary.main }}
      >
        <Toolbar>
          <Typography
            variant="h5"
            noWrap
            component="div"
            sx={{ flexGrow: 1 }}
            color="white"
          >
            HACKER
            <span style={{ color: theme.palette.primary.main }}>STORY</span>
          </Typography>
          <Search>
            <SearchIconWrapper>
              <SearchIcon sx={{ color: "white" }} />
            </SearchIconWrapper>
            <StyledInputBase
              data-testid="styledInputBar"
              placeholder="Searchâ€¦"
              inputProps={{ "aria-label": "search" }}
              value={queryText}
              onChange={handleChange}
              sx={{ color: "#fff " }}
            />
          </Search>
        </Toolbar>
      </AppBar>
      <Box sx={{ width: "100%" }}>
        <LinearProgress sx={{ color: theme.palette.primary.main }} />
      </Box>
    </Box>
  );
};

export default SearchAppBar;
