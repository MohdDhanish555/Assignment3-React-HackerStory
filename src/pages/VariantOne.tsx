import { Box, CircularProgress, Pagination } from "@mui/material";
import { useContext, useEffect, useState } from "react";
import List from "../components/List";
import axios from "axios";
import { AppContext, API } from "../App";
import { useDebounce } from "../hooks/UseDebounce";

let totalPageCount = 0;

const VariantOne = () => {
  const [page, setPage] = useState(0);
  const [isLoading, setIsLoading] = useState(true);
  const [isError, setIsError] = useState(false);
  const [stories, setStories] = useState([]);
  const [queryText] = useContext(AppContext);

  useEffect(() => {
    setPage(0);
  }, [queryText]);

  const handleChange = (event: React.ChangeEvent<unknown>, value: number) => {
    setPage(value - 1);
  };

  const debouncedURL = useDebounce(`${API}query=${queryText}&page=${page}`);

  useEffect(() => {
    setIsLoading(true);
    setIsError(false);
    axios({
      method: "GET",
      url: debouncedURL,
    })
      .then((response) => {
        setStories(response.data.hits);
       totalPageCount = response.data.nbPages;
        setIsLoading(false);
      })
      .catch((error) => {
        setIsError(true);
      });
  }, [debouncedURL]);

  return (
    <>
      {isError ? (
        <h1>Error</h1>
      ) : (
        <Box display="flex" flexDirection="column" alignItems="center">
            <Pagination
            variant="outlined"
            color="primary"
            count={totalPageCount}
            page={page + 1}
            boundaryCount={15}
            onChange={handleChange}
            sx={{ mt: ".5rem" }}
          />
          {isLoading ? (
            <Box
              display="flex"
              justifyContent="center"
              alignItems="center"
              height="80vh"
            >
              <CircularProgress />
            </Box>
          ) : (
            <List stories={stories} setStories={setStories} />
          )}
        </Box>
      )}
    </>
  );
};

export default VariantOne;
