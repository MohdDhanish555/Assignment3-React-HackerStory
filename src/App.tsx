import SearchAppBar from "./components/SearchAppBar";
import { createContext, useState } from "react";
import { BrowserRouter, Route, Routes } from "react-router-dom";
import VariantOne from "./pages/VariantOne";
import VariantTwo from "./pages/VariantTwo";
import Home from "./pages/Home";
import RawContent from "./pages/RawContent";
import { ThemeProvider, createTheme } from "@mui/material/styles";

export const API = "https://hn.algolia.com/api/v1/search?";
export const AppContext: any = createContext([]);

declare module "@mui/material/styles" {
  interface Palette {
    tertiary: Palette["primary"];
  }
  interface PaletteOptions {
    tertiary: PaletteOptions["primary"];
  }
}

const theme = createTheme({
  palette: {
    primary: {
      main: "#9ACD32",
    },
    secondary: {
      main: "#262626",
    },
    tertiary: {
      main: "#00e6e6",
    },
  },
});

function App() {
  const [queryText, setQueryText] = useState("");

  return (
    <AppContext.Provider value={[queryText, setQueryText]}>
      <ThemeProvider theme={theme}>
        <BrowserRouter>
        <SearchAppBar />
        <Routes>
          <Route path="/" element={<Home />} />
          <Route path="/variantOne" element={<VariantOne />} />
          <Route path="/variantTwo" element={<VariantTwo />} />
          <Route path="/rawContent" element={<RawContent />} />
        </Routes>
        </BrowserRouter>
      </ThemeProvider>
    </AppContext.Provider>
  );
}

export default App;
