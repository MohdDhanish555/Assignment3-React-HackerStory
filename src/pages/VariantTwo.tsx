import { Box, CircularProgress } from "@mui/material";
import axios from "axios";
import { useCallback, useContext, useEffect, useRef, useState } from "react";
import { API, AppContext, StoryType } from "../App";
import List from "../components/List";
import { useDebounce } from "../hooks/UseDebounce";



const VariantTwo = () => {
  const [page, setPage] = useState(0);
  const [queryText] = useContext(AppContext);
  const [stories, setStories] = useState<StoryType[]>([]);
  const [isLoading, setIsLoading] = useState(true);
  const [isError, setIsError] = useState(false);
  const [hasMoreData, setHasMoreData] = useState(false);

  useEffect(() => {
    setStories([]);
    setPage(0);
  }, [queryText]);

  const debouncedURL = useDebounce(`${API}query=${queryText}&page=${page}`);

  useEffect(() => {
    setIsLoading(true);
    setIsError(false);
    axios({
      method: "GET",
      url: debouncedURL,
    })
      .then((response) => {
        setStories((stories: any) => {
          return [...new Set([...stories, ...response.data.hits])];
        });
        setHasMoreData(response.data.hits.length > 0);
        setIsLoading(false);
      })
      .catch((error) => {
        setIsError(true);
      });
  }, [debouncedURL]);

  const observer : any = useRef();
  const lastHitElement = useCallback(
    (node: any) => {
      if (isLoading) return;
      if (observer.current) observer.current.disconnect();
      observer.current = new IntersectionObserver(
        (entries) => {
          if (entries[0].isIntersecting && hasMoreData) {
            setPage((prevPage) => prevPage + 1);
          }
        },
        {
         rootMargin : "20px"
       }
      );
      if (node) observer.current.observe(node);
    },
    [isLoading, hasMoreData]
  );

  return (
    <>
      {isError ? (
        <h1>Error</h1>
      ) : (
        <Box
          display="flex"
          flexDirection="column"
          alignItems="center"
          justifyContent="center"
        >
          <List
            stories={stories}
            setStories={setStories}
            lastHitElement={lastHitElement}
          />
          {isLoading && (
            <Box mt="-50px">
              <CircularProgress />
            </Box>
          )}
        </Box>
      )}
    </>
  );
};

export default VariantTwo;
