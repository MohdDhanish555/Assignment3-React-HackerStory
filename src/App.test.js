import { fireEvent, getByTestId, render, screen } from "@testing-library/react";
import { BrowserRouter, useNavigate } from "react-router-dom";
import App, { AppContext } from "./App";
import List from "./components/List";
import SearchAppBar from "./components/SearchAppBar";
import Home from "./pages/Home";
import RawContent from "./pages/RawContent";
import VariantOne from "./pages/VariantOne";
import VariantTwo from "./pages/VariantTwo";


const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));




describe("SNAPSHOT testing", () => {
  test("", () => {
    const { AppComponent } = render(<App />);
    expect(AppComponent).toMatchSnapshot();
  });
});

const queryText = "react";
const setQueryText = jest.fn();

const storyOne = {
  title: "Learn React",
  url: "https://eprint.iacr.org/2021/1022",
  created_at: "2011-12-12",
  author: "grey-area",
  points: 1107,
  num_comments: 12,
  objectID: 1,
};

const storyTwo = {
  created_at: "2017-02-19T21:16:33.000Z",
  title: "Reflecting on one very, very strange year at Uber",
  url: "https://www.susanjfowler.com/blog/2017/2/19/reflecting-on-one-very-strange-year-at-uber",
  author: "grey-area",
  points: 4107,
  num_comments: 530,
  objectID: 3,
};

const storyThree = {
  created_at: "2021-04-05T14:04:22.000Z",
  title: "Google’s copying of the Java SE API was fair use [pdf]",
  url: "https://www.supremecourt.gov/opinions/20pdf/18-956_d18f.pdf",
  author: "pdoconnell",
  points: 4103,
  num_comments: 930,
  objectID: 4,
};

const stories = [storyOne, storyTwo];
const setStories = jest.fn();
const lastHitElement = storyOne;

describe("COMPONENTS testing", () => {
  test("APP Rendering...", () => {
    render(<App />);
  });

  test("HOME Rendering...", () => {
    render(
      <BrowserRouter>
        <Home />
      </BrowserRouter>
    );
    const btnV1 = screen.getByTestId("btn-V1");
    const btnV2 = screen.getByTestId("btn-V2");
    fireEvent.click(btnV1);
    fireEvent.click(btnV2);
    expect(mockedUsedNavigate).toHaveBeenCalledWith("/VariantOne");
    expect(mockedUsedNavigate).toHaveBeenCalledWith("/VariantTwo");
  });

  test("RAW CONTENT Rendering...", () => {
    render(
      <BrowserRouter>
        <RawContent />
      </BrowserRouter>
    );
    const btnRaw = screen.getByTestId("btn-Raw");
    fireEvent.click(btnRaw);
    expect(mockedUsedNavigate).toHaveBeenCalledWith(-1);
  });

  test("VARIANT ONE Rendering...", () => {
    const wrapper = ({ children }) => (
      <AppContext.Provider value={[queryText]}>{children}</AppContext.Provider>
    );
    render(
      <BrowserRouter>
        <VariantOne />
      </BrowserRouter>,
      { wrapper }
    );
    
  });

  test("VARIANT TWO Rendering...", () => {
    const wrapper = ({ children }) => (
      <AppContext.Provider value={[queryText]}>{children}</AppContext.Provider>
    );
    render(
      <BrowserRouter>
        <VariantTwo />
      </BrowserRouter>,
      { wrapper }
    );
  });

  test("SEARCH APPBAR Rendering...", () => {
    const wrapper = ({ children }) => (
      <AppContext.Provider value={[queryText, setQueryText]}>
        {children}
      </AppContext.Provider>
    );
    render(<SearchAppBar />, { wrapper });
    const inputBox = screen.getByPlaceholderText("Search…");
    fireEvent.change(inputBox, { target: { value: "TESTING" } });
  });

  test("LIST Rendering...", () => {
    render(
      <BrowserRouter>
        <List
          stories={stories}
          setStories={setStories}
          lastHitElement={lastHitElement}
        />
      </BrowserRouter>
    );
    const cellTitle1 = screen.getByTestId("cell-title1");
    const cellTitle2 = screen.getByTestId("cell-title2");
    const cellDelete1 = screen.getByTestId("cell-delete1");
    const cellDelete2 = screen.getByTestId("cell-delete2");
    fireEvent.click(cellTitle1);
    fireEvent.click(cellTitle2);
    fireEvent.click(cellDelete1);
    fireEvent.click(cellDelete2);
  });
});
