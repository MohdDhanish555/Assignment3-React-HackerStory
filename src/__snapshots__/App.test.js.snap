// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SNAPSHOT testing  1`] = `undefined`;
